### ISSUE TITLE
Feature request: sh to run toys if external command with than tame is not found

### ISSUE THREAD

**Issue by vt-alt** [landley/toybox#447]
Currently compiled in toys considered to be non-builtins and if there is no toy binary (symlink) they don't run. Sometimes it's useful (easier or possible) to run just single toybox/toysh binary without preparing symlink tree.

So idea is if external executable is not found but there is a toy with that name — run the toy. Benefit is we don't need the step of pre-symlinking cycle and just single toybox binary will be enough.

**Comment by landley** [comment]
Disable the TOYBOX_NORECURSE symbol in your config.

**Comment by vt-alt** [comment]
I have it unset already. (Just to prove I am not mistaking anything):
```
builder@x86_64:~/RPM/BUILD/toybox-0.8.10$ make
...
builder@x86_64:~/RPM/BUILD/toybox-0.8.10$ grep TOYBOX_NORECURSE .config
# CONFIG_TOYBOX_NORECURSE is not set
builder@x86_64:~/RPM/BUILD/toybox-0.8.10$ brctl
-bash: brctl: command not found
builder@x86_64:~/RPM/BUILD/toybox-0.8.10$ ./toybox brctl
brctl: Needs 1 argument (see "brctl --help")
builder@x86_64:~/RPM/BUILD/toybox-0.8.10$ ./toybox toysh
$ brctl
sh: brctl: No such file or directory
```
With this example, the idea is when `brctl` is invoked inside of `toysh` it would run `brctl` toy (when no `brctl` command exists).

**Comment by landley** [comment]
Hmmm, seems to have been a regression. Looking into it...

**Comment by landley** [comment]
Try commit aa6821ce1446

**Comment by landley** [comment]
Yeah, that's a design problem. Back when I maintained busybox there was a config option for whether the builtins should go before or after the $PATH, but that's ugly and I dowanna.

I suppose I could add an element to the $PATH to say where the builtins go? Gonna be in-band signaling wherever you put it, and we don't want to confuse other users of $PATH so magic symbols are out, but maybe I could declare a /toybox directory to mean "look inside toybox" instead of checking the filesystem? (Hey, if bash can have /dev/tcp and friends...)

So you could PATH=/toybox:$PATH or PATH=$PATH:/toybox and it would run the builtins? Then I don't need the "disable recursive execution" config option because it would be specified by the $PATH: if you don't put /toybox in the $PATH then it won't recurse.

I'd want Elliott's opinion on that though, because spurious opendir("/toybox") calls could trigger selinux monitoring weirdness or something. Toybox itself wouldn't do it, but things that inherited that $PATH string would. Then again, android doesn't like toybox to recurse, so...)

**Comment by landley** [comment]
Also such a feature can't _quite_ be reliable with enforced recursion limits, because despite me asking for execve(NULL, argv, envp) over fifteen years ago (https://lkml.iu.edu/hypermail/linux/kernel/0612.3/0238.html) the kernel guys never bothered to give a process an easy way to re-exec itself. The closest you can get is to exec /proc/self/exe which doesn't work if /proc isn't mounted and chroot easily screws up (despite which they had to do funky things to nerf it for containers...?) I could reclaim stack with longjmp() but that would leak memory and filehandles and so on so not really an improvement. Still... it should mostly work?

(Stares at tar looking for compressors via find_in_path(), and oneit.c being cute, and how to redo the shell search stuff _again_. But I _think_ that's it? The rest is just xexec() doing what it already does slightly differently...)

**Comment by vt-alt** [comment]
My thoughts: If there is no `/proc/self/exe` (I agree we cannot rely on its presence) you still can use `argv[0]` or `getauxval(AT_EXECFN)` (they could be different values[1]). name it `argv[0]` should be in `PATH`, and `AT_EXECFN` is a (relative/absolute) path (which should be directly re-execve-cutable). If files are deleted or overlaid by mounts — this maybe OK (because user perhaps wants that intentionally).

[1]

```
/tmp$ cat a.c
#include <stdio.h>
#include <sys/auxv.h>

int main(int argc, char **argv)
{
    printf("%s\n", argv[0]);
    printf("%s\n", (char *)getauxval(AT_EXECFN));
    return 0;
}
/tmp$ bash -c 'exec -a hello ./a'
hello
/tmp/a
```

**Comment by landley** [comment]
My test case is "run in empty chroot" (or as rescue binary on a filesystem with a broken libc or similar), so there _is_ no filename we're guaranteed to be able to run. Which unfortunately rules out exec() for toybox builtins, due to the lack of a re-exec-self system call that doesn't take a filename. And exec() does a lot of process resetting (freeing memory, closing filehandles, and so on) which toybox normally relies on.

It's _probably_ ok? I can have toybox fork() and recurse into a new command_main() function, which won't clear debris from the previous command but there shouldn't be that much. It's ok if the stack of commands doesn't get too deep, or if the calling program didn't accumulate too much debris. On a modern multi-gigabyte ram system it's probably fine, and mostly the zygote process is going to be "command shell" which has to clean up after itself in order for shell scripts to be long-lived...
